src/metabase/driver/mysql/actions.clj

```diff
@@ -52,67 +55,55 @@
-             (re-find #"Incorrect (.+?) value: '(.+)' for column (?:(.+)\.)??(?:(.+)\.)?(.+) at row (\d+)" error-message)]
-    [(cond-> {:message       (tru "incorrect value: {0}" value)
-              :column        (-> column
-                                 (str/replace #"^'(.*)'$" "$1")
-                                 remove-backticks)
-              :expected-type expected-type
-              :value         value
-              :row           (parse-long row)}
-       table    (assoc :table (remove-backticks table))
-       database (assoc :database (remove-backticks database)))]))
+(defmethod sql-jdbc.actions/maybe-parse-sql-error [:mysql actions.error/incorrect-value-type]
+  [_driver error-type _database error-message]
+  (when-let [[_ expected-type _value _database _table column _row]
+             (re-find #"Incorrect (.+?) value: '(.+)' for column (?:(.+)\.)??(?:(.+)\.)?(.+) at row (\d+)"  error-message)]
+    (let [column (-> column (str/replace #"^'(.*)'$" "$1") remove-backticks)]
+      {:type    error-type
+       :message (tru "Value for column {0} should be of type {1}" column expected-type)
+       :errors  {column (tru "The value should be of type {0}" expected-type)}})))
 
 (comment
-  (maybe-parse-fk-constraint-error nil "Cannot delete or update a parent row: a foreign key constraint fails (`food`.`y`, CONSTRAINT `y_ibfk_1` FOREIGN KEY (`x_id1`, `x_id2`) REFERENCES `x` (`id1`, `id2`))")
-  (maybe-parse-unique-constraint-error {:id 480} "(conn=10) Duplicate entry 'ID' for key 'string_pk.PRIMARY'")
-  (maybe-parse-not-null-error nil "Column 'f1' cannot be null")
-  (maybe-parse-incorrect-type nil "Incorrect integer value: 'not boolean' for column `G__168815`.`types`.`boolean` at row 1")
-  (maybe-parse-incorrect-type nil "(conn=183) Incorrect integer value: 'STRING' for column 'id' at row 1")
-  (maybe-parse-incorrect-type nil "(conn=183) Incorrect integer value: 'STRING' for column `table`.`id` at row 1")
+  (sql-jdbc.actions/maybe-parse-sql-error :mysql actions.error/violate-foreign-key-constraint nil
+                                          "(conn=21) Cannot delete or update a parent row: a foreign key constraint fails (`action-error-handling`.`user`, CONSTRAINT `user_group-id_group_-159406530` FOREIGN KEY (`group-id`) REFERENCES `group` (`id`))")
+  (sql-jdbc.actions/maybe-parse-sql-error :mysql actions.error/violate-unique-constraint {:id 3}
+                                          "(conn=10) Duplicate entry 'ID' for key 'string_pk.PRIMARY'")
+  (sql-jdbc.actions/maybe-parse-sql-error :mysql actions.error/violate-not-null-constraint nil
+                                          "Column 'f1' cannot be null")
+  (sql-jdbc.actions/maybe-parse-sql-error :mysql actions.error/incorrect-value-type nil
+                                          "(conn=183) Incorrect integer value: 'STRING' for column `table`.`id` at row 1")
   nil)
 
-(defmethod sql-jdbc.actions/parse-sql-error :mysql
```

Comment:
removing `sql-jdbc.actions/parse-sql-error` was my original motivation, I think having a list of functions for each driver like this is hard to know which drivers support which parsing error function.